From 9af092b57cc10bf409b2a69d17ebf2bf96754955 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 12 Sep 2024 19:19:41 +0200
Subject: [PATCH 1201/1695] drivers: stm32_cpu_opp: manage OPP for all SoC

The OPP must be restored after Standby mode for all STM32 SoCs
when SCMI stack (the OP-TEE stack or SCP stack) is used for DVFS
(including STM32MP15 and STM32MP25) and not only for STM32MP13.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git dabaf1366cca08a1175cba3f3d409b9a8b7df522 drivers: stm32_cpu_opp: manage OPP for all SoC]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Idc13cb03b5460bbf7426fb945747fef1fb6753ab
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/404217
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_cpu_opp.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/core/drivers/stm32_cpu_opp.c b/core/drivers/stm32_cpu_opp.c
index c94196632..311448af1 100644
--- a/core/drivers/stm32_cpu_opp.c
+++ b/core/drivers/stm32_cpu_opp.c
@@ -518,7 +518,6 @@ TEE_Result stm32_cpu_opp_read_level(unsigned int *level)
 }
 #endif /*CFG_SCPFW_MOD_DVFS*/
 
-#ifdef CFG_STM32MP13
 static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint __unused,
 			     const struct pm_callback_handle *hdl __unused)
 {
@@ -527,11 +526,20 @@ static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint __unused,
 
 	assert(op == PM_OP_SUSPEND || op == PM_OP_RESUME);
 
+	/* nothing to do if OPP is managed by Linux and not by SCMI */
+	if (!IS_ENABLED(CFG_SCMI_MSG_PERF_DOMAIN) &&
+	    !IS_ENABLED(CFG_SCPFW_MOD_DVFS))
+		return TEE_SUCCESS;
+
+	/* nothing to do if RCC clock tree is not lost */
+	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT))
+		return TEE_SUCCESS;
+
 #ifdef CFG_SCPFW_MOD_DVFS
 	if (op == PM_OP_SUSPEND) {
 		/*
 		 * When CFG_SCPFW_MOD_DVFS is enabled, save CPU OPP on suspend
-		 * for restoration at resmue. If CPU is not in an expected
+		 * for restoration at resume. If CPU is not in an expected
 		 * OPP state, fallback to default OPP at PM resume.
 		 */
 		unsigned int cur_khz = 0;
@@ -583,7 +591,6 @@ static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint __unused,
 	return TEE_SUCCESS;
 }
 DECLARE_KEEP_PAGER_PM(cpu_opp_pm);
-#endif
 
 static TEE_Result stm32_cpu_opp_is_supported(const void *fdt, int subnode)
 {
@@ -704,9 +711,7 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 	if (res)
 		return res;
 
-#ifdef CFG_STM32MP13
 	register_pm_driver_cb(cpu_opp_pm, NULL, "cpu-opp");
-#endif
 
 	return TEE_SUCCESS;
 }
-- 
2.47.2

